File: 29Items-Projects/7Algorithms/6-Backtracking/java/src/nqueens/NQueens.java
package nqueens;

import java.util.*;

public class NQueens {

    private final int n;
    private final List<int[]> solutions = new ArrayList<>();

    public NQueens(int n) {
        if (n <= 0) throw new IllegalArgumentException("Board size must be > 0");
        this.n = n;
    }

    public void solveAndPrint() {
        solve(new int[n], 0);
        System.out.printf("Found %d solution(s) for n=%d%n%n", solutions.size(), n);

        int limit = Math.min(solutions.size(), 4);
        for (int i = 0; i < limit; i++) {
            System.out.printf("Solution #%d:%n", i + 1);
            printBoard(solutions.get(i));
            System.out.println();
        }
        if (solutions.size() > limit) {
            System.out.printf("... and %d more solutions.%n", solutions.size() - limit);
        }
    }

    private void solve(int[] queens, int row) {
        if (row == n) {
            solutions.add(queens.clone());
            return;
        }
        for (int col = 0; col < n; col++) {
            if (isSafe(queens, row, col)) {
                queens[row] = col;
                solve(queens, row + 1);
            }
        }
    }

    private boolean isSafe(int[] queens, int row, int col) {
        for (int r = 0; r < row; r++) {
            if (queens[r] == col || Math.abs(queens[r] - col) == row - r) {
                return false;
            }
        }
        return true;
    }

    private void printBoard(int[] queens) {
        for (int r = 0; r < n; r++) {
            StringBuilder sb = new StringBuilder("|");
            for (int c = 0; c < n; c++) {
                sb.append(queens[r] == c ? " Q " : " . ");
            }
            sb.append("|");
            System.out.println(sb);
        }
    }
}

File: 29Items-Projects/7Algorithms/6-Backtracking/python/nqueens.py
"""N-Queens — place n queens on an n×n board with no mutual attacks."""


class NQueens:
    def __init__(self, n: int):
        if n <= 0:
            raise ValueError("Board size must be > 0")
        self.n = n
        self.solutions: list[list[int]] = []

    def solve_and_print(self) -> None:
        self._solve([], 0)
        print(f"Found {len(self.solutions)} solution(s) for n={self.n}\n")

        limit = min(len(self.solutions), 4)
        for i, sol in enumerate(self.solutions[:limit], 1):
            print(f"Solution #{i}:")
            self._print_board(sol)
            print()

        if len(self.solutions) > limit:
            print(f"... and {len(self.solutions) - limit} more solutions.")

    def _solve(self, queens: list[int], row: int) -> None:
        if row == self.n:
            self.solutions.append(queens[:])
            return
        for col in range(self.n):
            if self._is_safe(queens, row, col):
                queens.append(col)
                self._solve(queens, row + 1)
                queens.pop()

    def _is_safe(self, queens: list[int], row: int, col: int) -> bool:
        for r, c in enumerate(queens):
            if c == col or abs(c - col) == row - r:
                return False
        return True

    def _print_board(self, queens: list[int]) -> None:
        for r in range(self.n):
            row = "│"
            for c in range(self.n):
                row += " ♛ " if queens[r] == c else " · "
            row += "│"
            print(row)
